client/my-sites/plugins/plugins-browser/index.jsx

```diff
@@ -18,92 +18,64 @@ import NavTabs from 'components/section-nav/tabs';
 				search_query: this.props.search
 			} );
 		}
 	},
 
-	getInitialState() {
-		return this.getPluginsLists( this.props.search );
-	},
-
-	componentWillUnmount() {
-		PluginsListStore.removeListener( 'change', this.refreshLists );
-		this.props.sites.removeListener( 'change', this.refreshLists );
-	},
-
-	componentWillReceiveProps( newProps ) {
-		this.refreshLists( newProps.search );
-	},
-
-	refreshLists( search ) {
-		this.setState( this.getPluginsLists( search || this.props.search ) );
-	},
-
-	fetchNextPagePlugins() {
-		let doSearch = true;
-
-		if ( this.state.fullLists.search && this.state.fullLists.search.fetching ) {
-			doSearch = false;
-		}
-
-		if ( this.state.fullLists.search && this.state.fullLists.search.list && this.state.fullLists.search.list.length < 10 ) {
-			doSearch = false;
-		}
-
-		if ( this.props.search && doSearch ) {
-			PluginsActions.fetchNextCategoryPage( 'search', this.props.search );
-		} else if ( this.props.category ) {
-			PluginsActions.fetchNextCategoryPage( this.props.category );
+	componentWillReceiveProps( nextProps ) {
+		if ( nextProps.search ) {
+			if ( nextProps.search !== this.props.search ) {
+				this.fetchNextPagePlugins( nextProps.search );
+			}
 		}
-	},
-
-	getPluginsLists( search ) {
-		const shortLists = {},
-			fullLists = {};
-		this.visibleCategories.forEach( category => {
-			shortLists[ category ] = PluginsListStore.getShortList( category );
-			fullLists[ category ] = PluginsListStore.getFullList( category );
+		this.setState( {
+			accessError: pluginsAccessControl.hasRestrictedAccess()
 		} );
-		fullLists.search = PluginsListStore.getSearchList( search );
-		return {
-			accessError: pluginsAccessControl.hasRestrictedAccess(),
-			shortLists: shortLists,
-			fullLists: fullLists
-		};
-	},
-
-	getPluginsShortList( listName ) {
-		return this.state.shortLists[ listName ] ? this.state.shortLists[ listName ].list : [];
 	},
 
-	getPluginsFullList( listName ) {
-		return this.state.fullLists[ listName ] ? this.state.fullLists[ listName ].list : [];
+	fetchNextPagePlugins( searchTerm ) {
+		searchTerm = typeof searchTerm === 'string' ? searchTerm : this.props.search;
+		if ( searchTerm ) {
+			const lastFetchedPage = this.props.lastPage.search >= FIST_PAGE ? this.props.lastPage.search : -1;
+			if ( this.props.currentSearchTerm !== searchTerm ) {
+				this.props.fetchPluginsList( 'search', FIST_PAGE, searchTerm );
+			} else if ( this.props.canFetchList( 'search', lastFetchedPage + 1, searchTerm ) ) {
+				this.props.fetchPluginsList( 'search', lastFetchedPage + 1, searchTerm );
+			}
+		} else if ( this.props.category ) {
+			const lastFetchedPage = this.props.lastPage[ this.props.category ] >= FIST_PAGE
+				? this.props.lastPage[ this.props.category ]
+				: -1;
+			if ( this.props.canFetchList( this.props.category, lastFetchedPage + 1 ) ) {
+				this.props.fetchPluginsList( this.props.category, lastFetchedPage + 1 );
+			}
+		}
```

Comment:
something about this is hard for me to parse. there is much nesting going on and I'm not sure the sequence of what should be happening is clear. could we flatten this down maybe?

```js
const {
	canFetchList,
	category,
	currentSearchTerm,
	fetchPluginList,
	lastPage,
} = this.props;

if ( searchTerm && currentSearchTerm !== searchTerm ) {
	return fetchPluginList( 'search', FIRST_PAGE, searchTerm );
}

// note we are coercing the boolean here to zero
// out the function when less than FIRST PAGE
const nextSearchPage = ( lastPage.search + 1 ) * ( lastPage.search >= FIRST_PAGE );
if ( searchTerm && canFetchList( 'search', nextSearchPage, searchTerm ) ) {
	return fetchPluginList( 'search', nextSearchPage, searchTerm );
}

const nextCategoryPage = ( lastPage[ category ] + 1 ) * ( lastPage[ category ] >= FIRST_PAGE );
if ( category && canFetchList( category, nextCategoryPage ) ) {
	return fetchPluginsList( category, nextCategoryPage );
}
```